kbase-extension/static/kbase/js/widgets/narrative_core/kbaseNarrativeDataList.js

```diff
@@ -573,34 +585,76 @@ define([
                 }.bind(this));
             }.bind(this);
 
-            return Promise.resolve(
-                this.serviceClient.sync_call(
-                    'NarrativeService.list_objects_with_sets', [{
-                        'ws_name': this.ws_name,
-                        'includeMetadata': 1
-                    }]
-                )
-            )
-                .then(function (result) {
-                    result = result[0]['data'];
-
-                    for (var i = 0; i < result.length; i++) {
-                        var obj = result[i];
+            var narrativeService = new DynamicServiceClient({
+                module: 'NarrativeService',
+                url: Config.url('service_wizard'),
+                token: this.token
+            });
+
+            var workspace = new GenericClient({
+                module: 'Workspace',
+                url: Config.url('workspace'),
+                token: this.token
+            });
+
+            var reportRegex = /^KBaseReport.Report-/;
+
+            return narrativeService.callFunc('list_objects_with_sets', [{
```

Comment:
What does this do, exactly? It looks like you're stitching in to the Data Panel populator and looking up everything that references each object, and filtering for reports? That should either just be part of the NarrativeService call. Or, to make it faster, just find the report when the it's requested. In a large narrative, this might take forever.

Also, copied objects (i.e. everything in a copied narrative) lose that link. You can find it by tracing back up the copy chain, but that's also something that shouldn't be applied to every object in a Narrative, and only on request.

Reply:
I'm working on a function in NarrativeService right now that should do what you need.